app/src/common/shared/org/mozilla/vrbrowser/ui/CustomKeyboard.java

```diff
@@ -25,7 +26,132 @@
     public static final int KEYCODE_STRING_COM = -12;
 
     public CustomKeyboard(Context context, int xmlLayoutResId) {
-        super(context, xmlLayoutResId);
+        super(context, xmlLayoutResId, 0);
+    }
+
+
+    public CustomKeyboard (Context context, int layoutTemplateResId, CharSequence characters, int columns, int horizontalPadding) {
+        this(context, layoutTemplateResId);
+
+        int x = 0;
+        int y = 0;
+        int column = 0;
+        setParentField(this, "mTotalWidth", 0);
+
+        int mDefaultWidth = getParentFieldInt(this, "mDefaultWidth");
+        int mDefaultHeight = getParentFieldInt(this, "mDefaultHeight");
+        int mDefaultHorizontalGap = getParentFieldInt(this, "mDefaultHorizontalGap");
+        int mDefaultVerticalGap = getParentFieldInt(this, "mDefaultVerticalGap");
+        int mDisplayWidth = getParentFieldInt(this, "mDisplayWidth");
+        ArrayList<Row> rows = (ArrayList<Row>)getParentFieldObject(this, "rows");
+
+        int rowsNum = (int)Math.ceil((characters.length()+0.1)/columns);
+        final int maxColumns = columns == -1 ? Integer.MAX_VALUE : columns;
+        Row[] mRows = new Row[maxColumns];
+        for (int i=0; i<rowsNum; i++) {
+            Row row = new Row(this);
+            row.defaultHeight = mDefaultHeight;
+            row.defaultWidth = mDefaultWidth;
+            row.defaultHorizontalGap = mDefaultHorizontalGap;
+            row.verticalGap = mDefaultVerticalGap;
+            row.rowEdgeFlags = EDGE_TOP;
+            mRows[i] = row;
+        }
+        for (int i = 0; i < characters.length(); i++) {
+            char c = characters.charAt(i);
+            int rowIndex = (int)Math.floor(i/columns);
+
+            final Key key = new Key(mRows[rowIndex]);
+
+            if (column >= maxColumns
+                    || x + mDefaultWidth + horizontalPadding > mDisplayWidth) {
+
+                mRows[rowIndex].rowEdgeFlags = EDGE_BOTTOM;
+
+                x = 0;
+                y += mDefaultVerticalGap + mDefaultHeight;
+                column = 0;
+            }
+
+            key.x = x;
+            key.y = y;
+            key.label = String.valueOf(c);
+            key.codes = new int[]{c};
+            column++;
+            x += key.width + key.gap;
+            getKeys().add(key);
+            ArrayList<Key> mKeys = (ArrayList<Key>) getFieldObject(mRows[rowIndex], "mKeys");
```

Comment:
This could crash if getFieldObject returns the 0 Integer (Also if we update it to return null). Add some checks to avoid runtime crashes if the fields are not found in some Android OS version